### PR TITLE
Integrations service for CRUD operations

### PR DESCRIPTION
**Pull request by marcoandredinis**
This adds the services that will handle the CRUD operations for the Integration resource.
It also includes a custom JSON serializer.

Part of RFD https://github.com/gravitational/teleport/pull/22556

### CODE REVIEW

File: api/types/integration.go [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+
+// Integration specifies is a connection configuration between Teleport and a 3rd party system.
+type Integration interface {
+	ResourceWithLabels
+
+	// GetStatusCode returns the current integration status as a code.
+	GetStatusCode() IntegrationSpecV1_IntegrationStatus
+
+	// GetStatus returns the integration status as string.
+	GetStatus() string
```
[the remark this answers is not in the document]
Comment (marcoandredinis) [reply api/types/integration.go:41]: I think we can drop the `GetStatusCode()` :+1:

File: api/types/integration.go [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+// Spec.SubKindSpec is expecting the `SubKindSpec` json tag, however we are using snake_case everywhere.
+// So, we create a local type that has the expected json tag (`sub_kind_spec`) and use it to unmarshal and then copy
+// to the proper type.
+func (ig *IntegrationV1) UnmarshalJSON(data []byte) error {
+	if ig == nil {
+		ig = &IntegrationV1{}
+	}
+
+	d := struct {
+		ResourceHeader `json:""`
```
[the remark this answers is not in the document]
Comment (marcoandredinis) [reply api/types/integration.go:215]: Not adding them means we have the `ResourceHeader` fields at the root level in the json output.
Which seems more aligned with the other resources representation (either in json or yaml)

Example
without JSON Tag:
```json
{
    "kind": "integration",
    "sub_kind": "aws-oidc",
    "version": "v1",
    "metadata": {
        "name": "some-integration"
    },
    "spec": {
        "status": 1,
        "subkind_spec": {
            "aws_oidc": {
                "role_arn": "arn:aws:iam::123456789012:role/DevTeams"
            }
        }
    }
}
```

with JSON Tag `ResourceHeader 'json:"header"'`:
```json
{
    "header": {
        "kind": "integration",
        "sub_kind": "aws-oidc",
        "version": "v1",
        "metadata": {
            "name": "some-integration"
        }
    },
    "spec": {
        "status": 1,
        "subkind_spec": {
            "aws_oidc": {
                "role_arn": "arn:aws:iam::123456789012:role/DevTeams"
            }
        }
    }
}
```

**Review by rosstimothy — APPROVED** [review]

File: lib/auth/integration/integrationv1/service.go [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+	Cache      services.IntegrationsGetter
+	Backend    services.Integrations
+	Logger     *logrus.Entry
+}
+
+// Service implements the teleport.integration.v1.IntegrationService RPC service.
+type Service struct {
+	integrationpb.UnimplementedIntegrationServiceServer
+	authorizer authz.Authorizer
+	cache      services.IntegrationsGetter
```
[the remark this answers is not in the document]
Comment (marcoandredinis) [reply lib/auth/integration/integrationv1/service.go:45]: I'll backport this change to v12 and cache the Integration resource only for Auth Service.

For v13 (master), I'll add cache for both Auth and Proxy services.
Does it look alright?

File: api/types/integration.go [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+	ig.Spec.SubKindSpec = &IntegrationSpecV1_AWSOIDC{
+		AWSOIDC: awsOIDCSpec,
+	}
+}
+
+// GetStatusCode returns the integration status as a code.
+// It can be one of:
+// - paused: integration was just configured or was disabled by the user
+// - running: integration is ready to do requests
+// - error: integration has an error and should
```
Comment (r0mant) [inline api/types/integration.go:150]: Unfinished phrase: "and should ..."

File: api/types/integration.go [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+func (ig *IntegrationV1) GetStatusCode() IntegrationSpecV1_IntegrationStatus {
+	return ig.Spec.Status
+}
+
+// GetStatus returns the integration status.
+// It can be one of:
+// - paused: integration was just configured or was disabled by the user
+// - running: integration is ready to do requests
+// - error: integration has an error and should
```
Comment (r0mant) [inline api/types/integration.go:159]: Unfinished phrase: "and should ..."

File: lib/auth/integration/integrationv1/service.go [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+	switch {
+	case cfg.Cache == nil:
+		return nil, trace.BadParameter("cache is required")
+	case cfg.Backend == nil:
+		return nil, trace.BadParameter("backend is required")
+	case cfg.Authorizer == nil:
+		return nil, trace.BadParameter("authorizer is required")
+	case cfg.Logger == nil:
+		cfg.Logger = logrus.WithField(trace.Component, "integrations.service")
+	}
```
Comment (r0mant) [inline lib/auth/integration/integrationv1/service.go:61]: Why not do `CheckAndSetDefaults()` method on the config as we usually do?

File: api/types/integration.go [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+// SetAWSOIDCIntegrationSpec sets the specific fields for the `aws-oidc` subkind integration.
+func (ig *IntegrationV1) SetAWSOIDCIntegrationSpec(awsOIDCSpec *AWSOIDCIntegrationSpecV1) {
+	ig.Spec.SubKindSpec = &IntegrationSpecV1_AWSOIDC{
+		AWSOIDC: awsOIDCSpec,
+	}
+}
+
+// GetStatusCode returns the integration status as a code.
+// It can be one of:
+// - paused: integration was just configured or was disabled by the user
```
Comment (r0mant) [inline api/types/integration.go:148]: What's the use-case for paused vs running vs error state?

File: lib/auth/integration/integrationv1/service.go (hunk 6 above)
Comment (marcoandredinis) [reply lib/auth/integration/integrationv1/service.go:61]: Sorry, copy paste from another service
Fixed :+1:

File: api/types/integration.go (hunk 7 above)
Comment (marcoandredinis) [reply api/types/integration.go:148]: I'll get back to this comment after resolving the comment in the RFD PR
https://github.com/gravitational/teleport/pull/22556#discussion_r1157561311

Comment (r0mant) [reply api/types/integration.go:148]: Should this now be removed?

Comment (marcoandredinis) [reply api/types/integration.go]: Yes, just remove all the Integration.Status reference :+1:

**Review by r0mant — APPROVED** [review]

**Pull request merged**